File: benchmark/benchmark.nim
# import std/[math, algorithm, strutils, strformat, sequtils, tables]
# import ../src/[values]
# import nimprof
import ./src/nim/[common]
import ./src/nim/parazoa/arr as para_arr
import ./src/nim/parazoa/map as para_map
import ./src/nim/nim_persistent_vector/arr as pers_arr
import ./src/nim/vec as vec
import ./src/nim/map as map

const RUN_PARAZOA    = true
const RUN_PERSVECTOR = true
const RUN_VEC        = true
const RUN_MAP        = true
const RUN_SANITY     = true

const RUN_MAPS       = true
const RUN_ARRS       = true

proc sanity_check(tr: TaskResult, sz, n: int) =
  var s = 0.0
  let Start = get_time()
  for i in 0..<n:
    s += i.float64
    if tr.runs.len > 1000000: echo s
    if tr.runs.len > 10000000: echo s
  tr.add(get_time() - Start)

proc output_results() =
  write_row("\"key\",\"sys\",\"desc\",\"runs\",\"minimum\",\"maximum\",\"mean\",\"median\"")
  for tr in csv_rows:
    tr.to_row.write_row

const VEC        = "vec"
const MAP        = "map"
const PARAZOA    = "parazoa"
const PERSVECTOR = "persvector"
const IMPERATIVE = "imperative"

proc run_benchmarks() =
  warmup()
  if RUN_SANITY:
    bench("sanity_check", "--", sanity_check, 0, 5000000)
    bench("sanity_check", "--", sanity_check, 0, 50000)
    bench("sanity_check", "--", sanity_check, 0, 500)

  # value benchmarks
  block:
    for it in [10, 100, 1000]:
      if RUN_MAPS:
        if RUN_PARAZOA:
          bench("map_create", PARAZOA, parazoa_map_create, 0, it)
        if RUN_MAP:
          bench("map_create", MAP, pmap_create, 0, it)

      if RUN_ARRS:
        if RUN_VEC:
          bench("arr_create", VEC, pvec_arr_create, 0, it)
        if RUN_PERSVECTOR:
          bench("arr_create", PERSVECTOR, persvector_arr_create, 0, it)
        if RUN_PARAZOA:
          bench("arr_create", PARAZOA, parazoa_arr_create, 0, it)

      for sz in [1, 10, 100, 1000]:
        if it < 100 and sz < 100: continue
        if it > 100 and sz >= 100: continue
        if it >= 100 and sz > 100: continue
        # echo "it: ", it, " sz: ", sz

        if RUN_MAPS:
          if RUN_MAP:
            bench("map_add_entry", MAP, pmap_add_entry, sz, it)
            bench("map_add_entry_multiple", MAP, pmap_add_entry_multiple, sz, it)
            bench("map_overwrite_entry", MAP, pmap_overwrite_entry, sz, it)
            bench("map_del_entry", MAP, pmap_del_entry, sz, it)
            bench("map_merge", MAP, pmap_merge, sz, it)
            bench("map_has_key_true", MAP, pmap_has_key_true, sz, it)
            bench("map_has_key_false", MAP, pmap_has_key_false, sz, it)
            bench("map_get_existing", MAP, pmap_get_existing, sz, it)
            bench("map_get_non_existing", MAP, pmap_get_non_existing, sz, it)
            bench("map_iter_keys", MAP, pmap_iter_keys, sz, it)
            bench("map_iter_values", MAP, pmap_iter_values, sz, it)
            bench("map_iter_entries", MAP, pmap_iter_entries, sz, it)
            bench("map_equal_true", MAP, pmap_equal_true, sz, it)
            bench("map_equal_false", MAP, pmap_equal_false, sz, it)

          if RUN_PARAZOA:
            bench("map_add_entry", PARAZOA, parazoa_map_add_entry, sz, it)
            bench("map_add_entry_multiple", PARAZOA, parazoa_map_add_entry_multiple, sz, it)
            bench("map_overwrite_entry", PARAZOA, parazoa_map_overwrite_entry, sz, it)
            bench("map_del_entry", PARAZOA, parazoa_map_del_entry, sz, it)
            bench("map_merge", PARAZOA, parazoa_map_merge, sz, it)
            bench("map_has_key_true", PARAZOA, parazoa_map_has_key_true, sz, it)
            bench("map_has_key_false", PARAZOA, parazoa_map_has_key_false, sz, it)
            bench("map_get_existing", PARAZOA, parazoa_map_get_existing, sz, it)
            bench("map_get_non_existing", PARAZOA, parazoa_map_get_non_existing, sz, it)
            bench("map_iter_keys", PARAZOA, parazoa_map_iter_keys, sz, it)
            bench("map_iter_values", PARAZOA, parazoa_map_iter_values, sz, it)
            bench("map_iter_entries", PARAZOA, parazoa_map_iter_entries, sz, it)
            bench("map_equal_true", PARAZOA, parazoa_map_equal_true, sz, it)
            bench("map_equal_false", PARAZOA, parazoa_map_equal_false, sz, it)

        if RUN_ARRS:
          if RUN_VEC:
            bench("arr_push", VEC, pvec_arr_push, sz, it)
            bench("arr_pop", VEC, pvec_arr_pop, sz, it)
            bench("arr_get_existing", VEC, pvec_arr_get_existing, sz, it)
            bench("arr_get_non_existing", VEC, pvec_arr_get_non_existing, sz, it)
            bench("arr_set", VEC, pvec_arr_set, sz, it)
            bench("arr_iter", VEC, pvec_arr_iter, sz, it)
            bench("arr_equal_true", VEC, pvec_arr_equal_true, sz, it)
            bench("arr_equal_false", VEC, pvec_arr_equal_false, sz, it)

          if RUN_PERSVECTOR:
            bench("arr_push", PERSVECTOR, persvector_arr_push, sz, it)
            bench("arr_pop", PERSVECTOR, persvector_arr_pop, sz, it)
            bench("arr_get_existing", PERSVECTOR, persvector_arr_get_existing, sz, it)
            # bench("arr_get_non_existing", PERSVECTOR, persvector_arr_get_non_existing, sz, it)
            bench("arr_set", PERSVECTOR, persvector_arr_set, sz, it)
            bench("arr_iter", PERSVECTOR, persvector_arr_iter, sz, it)
            # bench("arr_equal_true", PERSVECTOR, persvector_arr_equal_true, sz, it)
            # bench("arr_equal_false", PERSVECTOR, persvector_arr_equal_false, sz, it)

          if RUN_PARAZOA:
            bench("arr_push", PARAZOA, parazoa_arr_push, sz, it)
            bench("arr_pop", PARAZOA, parazoa_arr_pop, sz, it)
            bench("arr_get_existing", PARAZOA, parazoa_arr_get_existing, sz, it)
            bench("arr_get_non_existing", PARAZOA, parazoa_arr_get_non_existing, sz, it)
            bench("arr_set", PARAZOA, parazoa_arr_set, sz, it)
            bench("arr_iter", PARAZOA, parazoa_arr_iter, sz, it)
            bench("arr_equal_true", PARAZOA, parazoa_arr_equal_true, sz, it)
            bench("arr_equal_false", PARAZOA, parazoa_arr_equal_false, sz, it)

run_benchmarks()
output_results()

File: benchmark/benchmark.js
import fs from "node:fs";
import {
  immutable_arr_create,
  immutable_arr_equal_false,
  immutable_arr_equal_true,
  immutable_arr_get_existing,
  immutable_arr_get_non_existing,
  immutable_arr_iter,
  immutable_arr_pop,
  immutable_arr_push,
  immutable_arr_set,
  immutable_arr_slice,
  plain_arr_create,
  plain_arr_equal_false,
  plain_arr_equal_true,
  plain_arr_get_existing,
  plain_arr_get_non_existing,
  plain_arr_iter,
  plain_arr_pop_by_mutation,
  plain_arr_pop_by_spread,
  plain_arr_push_by_mutation,
  plain_arr_push_by_spread,
  plain_arr_set_by_mutation,
  plain_arr_set_by_spread,
  plain_arr_slice,
} from "./src/js/arr.js";
import {
  LOW_TIMEOUT,
  OUTPUT_PATH,
  bench_async,
  bench_sync,
  csv_rows,
  get_time,
  to_row,
  warmup,
} from "./src/js/common.js";
import {
  immer_pojo_add_entry,
  immer_pojo_add_entry_multiple,
  immer_pojo_add_entry_multiple_batched,
  immer_pojo_del_entry,
  immer_pojo_merge,
  immer_pojo_overwrite_entry,
  immutable_map_add_entry,
  immutable_map_add_entry_multiple,
  immutable_map_add_entry_multiple_batched,
  immutable_map_create,
  immutable_map_del_entry,
  immutable_map_equal_false,
  immutable_map_equal_true,
  immutable_map_get_existing,
  immutable_map_get_non_existing,
  immutable_map_has_key_false,
  immutable_map_has_key_true,
  immutable_map_iter_entries,
  immutable_map_iter_keys,
  immutable_map_iter_values,
  immutable_map_merge,
  immutable_map_overwrite_entry,
  pojo_add_entry_by_mutation,
  pojo_add_entry_by_mutation_multiple,
  pojo_add_entry_by_spread,
  pojo_add_entry_by_spread_multiple,
  pojo_add_entry_by_spread_multiple_batched,
  pojo_create,
  pojo_del_entry_by_mutation,
  pojo_del_entry_by_spread,
  pojo_equal_false,
  pojo_equal_true,
  pojo_get_existing,
  pojo_get_non_existing,
  pojo_has_key_false,
  pojo_has_key_true,
  pojo_iter_entries,
  pojo_iter_keys,
  pojo_iter_values,
  pojo_merge_by_mutation,
  pojo_merge_by_spread,
  pojo_overwrite_entry,
  pojo_overwrite_entry_by_spread,
} from "./src/js/map.js";

function sanity_check(tr, _sz, n) {
  let start = get_time();
  var s = 0.0;
  for (let f = 0; f < n; f++) {
    s += f;
    // Add these lines to keep this from getting optimized away
    if (tr.runs.length > 1000000) console.log(s);
    if (tr.runs.length > 10000000) console.log(s);
  }
  tr.runs.push(get_time() - start);
}

/* descriptions */
const PLAIN = "plain";
const PLAIN_MUTATION = "plain_mutation";
const PLAIN_SPREAD = "plain_spread";
const IMMER_POJO = "immer_pojo";
const IMMUTABLEJS = "_immutable.js"; /* add a leading _ so it sorts first; we compare against it */

async function run_benchmarks() {
  await warmup();
  bench_sync("sanity_check", "--", sanity_check, 0, 5000000);
  bench_sync("sanity_check", "--", sanity_check, 0, 50000);
  bench_sync("sanity_check", "--", sanity_check, 0, 500);

  /* value benchmarks */
  {
    /* prettier-ignore */
    for (let it of [10, 100, 1000]) {
      /* array */
      bench_sync("arr_create", PLAIN, plain_arr_create, 0, it, LOW_TIMEOUT);
      bench_sync("arr_create", IMMUTABLEJS, immutable_arr_create, 0, it, LOW_TIMEOUT);
      /* map */
      bench_sync("map_create", PLAIN, pojo_create, 0, it, LOW_TIMEOUT);
      bench_sync("map_create", IMMUTABLEJS, immutable_map_create, 0, it, LOW_TIMEOUT);
      for (let sz of [1, 10, 100, 1000]) {
        if (it < 100 && sz < 100) continue;
        if (it > 100 && sz >= 100) continue;
        if (it >= 100 && sz > 100) continue;
        /* array */
        {
          bench_sync("arr_push", PLAIN_MUTATION, plain_arr_push_by_mutation, sz, it, LOW_TIMEOUT);
          bench_sync("arr_push", PLAIN_SPREAD, plain_arr_push_by_spread, sz, it, LOW_TIMEOUT);
          bench_sync("arr_push", IMMUTABLEJS, immutable_arr_push, sz, it, LOW_TIMEOUT);
          bench_sync("arr_pop", PLAIN_MUTATION, plain_arr_pop_by_mutation, sz, it, LOW_TIMEOUT);
          bench_sync("arr_pop", PLAIN_SPREAD, plain_arr_pop_by_spread, sz, it, LOW_TIMEOUT);
          bench_sync("arr_pop", IMMUTABLEJS, immutable_arr_pop, sz, it, LOW_TIMEOUT);
          bench_sync("arr_slice", PLAIN, plain_arr_slice, sz, it, LOW_TIMEOUT);
          bench_sync("arr_slice", IMMUTABLEJS, immutable_arr_slice, sz, it, LOW_TIMEOUT);
          bench_sync("arr_get_existing", PLAIN, plain_arr_get_existing, sz, it, LOW_TIMEOUT);
          bench_sync("arr_get_existing", IMMUTABLEJS, immutable_arr_get_existing, sz, it, LOW_TIMEOUT);
          bench_sync("arr_get_non_existing", PLAIN, plain_arr_get_non_existing, sz, it, LOW_TIMEOUT);
          bench_sync("arr_get_non_existing", IMMUTABLEJS, immutable_arr_get_non_existing, sz, it, LOW_TIMEOUT);
          bench_sync("arr_set", PLAIN_MUTATION, plain_arr_set_by_mutation, sz, it, LOW_TIMEOUT);
          bench_sync("arr_set", PLAIN_SPREAD, plain_arr_set_by_spread, sz, it, LOW_TIMEOUT);
          bench_sync("arr_set", IMMUTABLEJS, immutable_arr_set, sz, it, LOW_TIMEOUT);
          bench_sync("arr_iter", PLAIN, plain_arr_iter, sz, it, LOW_TIMEOUT);
          bench_sync("arr_iter", IMMUTABLEJS, immutable_arr_iter, sz, it, LOW_TIMEOUT);
          bench_sync("arr_equal_true", PLAIN, plain_arr_equal_true, sz, it, LOW_TIMEOUT);
          bench_sync("arr_equal_true", IMMUTABLEJS, immutable_arr_equal_true, sz, it, LOW_TIMEOUT);
          bench_sync("arr_equal_false", PLAIN, plain_arr_equal_false, sz, it, LOW_TIMEOUT);
          bench_sync("arr_equal_false", IMMUTABLEJS, immutable_arr_equal_false, sz, it, LOW_TIMEOUT);
        }
        /* map */
        {
          bench_sync("map_add_entry", PLAIN_MUTATION, pojo_add_entry_by_mutation, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry", PLAIN_SPREAD, pojo_add_entry_by_spread, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry", IMMUTABLEJS, immutable_map_add_entry, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry", IMMER_POJO, immer_pojo_add_entry, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple", PLAIN_MUTATION, pojo_add_entry_by_mutation_multiple, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple", PLAIN_SPREAD, pojo_add_entry_by_spread_multiple, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple", IMMUTABLEJS, immutable_map_add_entry_multiple, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple", IMMER_POJO, immer_pojo_add_entry_multiple, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple_batched", PLAIN_MUTATION, pojo_add_entry_by_mutation_multiple, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple_batched", PLAIN_SPREAD, pojo_add_entry_by_spread_multiple_batched, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple_batched", IMMUTABLEJS, immutable_map_add_entry_multiple_batched, sz, it, LOW_TIMEOUT);
          bench_sync("map_add_entry_multiple_batched", IMMER_POJO, immer_pojo_add_entry_multiple_batched, sz, it, LOW_TIMEOUT);
          bench_sync("map_overwrite_entry", PLAIN_MUTATION, pojo_overwrite_entry, sz, it, LOW_TIMEOUT);
          bench_sync("map_overwrite_entry", PLAIN_SPREAD, pojo_overwrite_entry_by_spread, sz, it, LOW_TIMEOUT);
          bench_sync("map_overwrite_entry", IMMUTABLEJS, immutable_map_overwrite_entry, sz, it, LOW_TIMEOUT);
          bench_sync("map_overwrite_entry", IMMER_POJO, immer_pojo_overwrite_entry, sz, it, LOW_TIMEOUT);
          bench_sync("map_del_entry", PLAIN_MUTATION, pojo_del_entry_by_mutation, sz, it, LOW_TIMEOUT)
          bench_sync("map_del_entry", PLAIN_SPREAD, pojo_del_entry_by_spread, sz, it, LOW_TIMEOUT)
          bench_sync("map_del_entry", IMMUTABLEJS, immutable_map_del_entry, sz, it, LOW_TIMEOUT)
          bench_sync("map_del_entry", IMMER_POJO, immer_pojo_del_entry, sz, it, LOW_TIMEOUT)
          bench_sync("map_merge", PLAIN_MUTATION, pojo_merge_by_mutation, sz, it, LOW_TIMEOUT)
          bench_sync("map_merge", PLAIN_SPREAD, pojo_merge_by_spread, sz, it, LOW_TIMEOUT)
          bench_sync("map_merge", IMMUTABLEJS, immutable_map_merge, sz, it, LOW_TIMEOUT)
          bench_sync("map_merge", IMMER_POJO, immer_pojo_merge, sz, it, LOW_TIMEOUT)
          bench_sync("map_has_key_true", PLAIN, pojo_has_key_true, sz, it, LOW_TIMEOUT)
          bench_sync("map_has_key_true", IMMUTABLEJS, immutable_map_has_key_true, sz, it, LOW_TIMEOUT)
          bench_sync("map_has_key_false", PLAIN, pojo_has_key_false, sz, it, LOW_TIMEOUT)
          bench_sync("map_has_key_false", IMMUTABLEJS, immutable_map_has_key_false, sz, it, LOW_TIMEOUT)
          bench_sync("map_get_existing", PLAIN, pojo_get_existing, sz, it, LOW_TIMEOUT)
          bench_sync("map_get_existing", IMMUTABLEJS, immutable_map_get_existing, sz, it, LOW_TIMEOUT)
          bench_sync("map_get_non_existing", PLAIN, pojo_get_non_existing, sz, it, LOW_TIMEOUT)
          bench_sync("map_get_non_existing", IMMUTABLEJS, immutable_map_get_non_existing, sz, it, LOW_TIMEOUT)
          bench_sync("map_iter_keys", PLAIN, pojo_iter_keys, sz, it, LOW_TIMEOUT)
          bench_sync("map_iter_keys", IMMUTABLEJS, immutable_map_iter_keys, sz, it, LOW_TIMEOUT)
          bench_sync("map_iter_values", PLAIN, pojo_iter_values, sz, it, LOW_TIMEOUT)
          bench_sync("map_iter_values", IMMUTABLEJS, immutable_map_iter_values, sz, it, LOW_TIMEOUT)
          bench_sync("map_iter_entries", PLAIN, pojo_iter_entries, sz, it, LOW_TIMEOUT)
          bench_sync("map_iter_entries", IMMUTABLEJS, immutable_map_iter_entries, sz, it, LOW_TIMEOUT)
          bench_sync("map_equal_true", PLAIN, pojo_equal_true, sz, it, LOW_TIMEOUT)
          bench_sync("map_equal_true", IMMUTABLEJS, immutable_map_equal_true, sz, it, LOW_TIMEOUT)
          bench_sync("map_equal_false", PLAIN, pojo_equal_false, sz, it, LOW_TIMEOUT)
          bench_sync("map_equal_false", IMMUTABLEJS, immutable_map_equal_false, sz, it, LOW_TIMEOUT)
        }
      }
    }
  }
}

run_benchmarks().then(() => {
  fs.writeFileSync(
    OUTPUT_PATH,
    '"key","sys","desc","runs","minimum","maximum","mean","median"\n' +
      csv_rows.map(to_row).join("\n")
  );
});
